### DISCUSSION TITLE
Are there any docs or guides on how to write automated tests?

### DISCUSSION THREAD

**Discussion by jtuchel** [discord-net/Discord.Net#2125] [Q&A] [answered]
Hey there, 
are there any guides on how to write tests? I wasn't able to find any docs targeting this topic.

As an example I think this question describes "my struggles" a little bit

https://stackoverflow.com/questions/70134407/how-to-inject-a-fake-implementation-of-a-class-providing-no-interface-and-no-vir

but it would be nice if there would be a general "best practises for testing" section :)

**Comment by MichaelHochriegl** [comment, marked as answer]
Only solution to properly unit test this lib is to go with the last suggestion in you stackoverflow post.
So you have to create an Interface with an Implementation that uses an injected DiscordSocketClient to it.

Wrap the calls from the DiscordSocketClient in your Interface methods. You can then mock your Interface in your tests.

**Comment by jtuchel** [comment]
@MichaelHochriegl thanks. So yes, I could create something like this

```cs
    public interface IDiscordClient
    {
        Task CreateSession();
        Task DestroySession();
    }

    public sealed class DiscordClient : IDiscordClient
    {
        private readonly DiscordSocketClient _discordSocketClient;

        public DiscordClient(DiscordSocketClient discordSocketClient)
        {
            _discordSocketClient = discordSocketClient;
        }
        
        public async Task CreateSession()
        {
            await _discordSocketClient.LoginAsync(TokenType.Bot, "read token from injected options");
            await _discordSocketClient.StartAsync();
        }

        public async Task DestroySession()
        {
            await _discordSocketClient.LogoutAsync();
            await _discordSocketClient.StopAsync();
        }
    }
```

but I'm struggling with 2 things here

- I still won't be able to test `DiscordClient` right?
- When working with users, channels, guilds, ... it will be a huge effort to "reimplement" those methods inside the wrapper class ... ? So yeah, sometimes the client returns interfaces instead of classes. But still, you have to "copy paste" some methods just to call the real method under the hood

Any suggestions?

**Comment by MichaelHochriegl** [comment]
If you want to test the actual DiscordClient you would need to do Integration Tests. Unit tests mock dependencies away. The DiscordClient is your dependency here, so for unit testing you make this mockable with the code you posted above.

You don't reimplement this mocked methods, you ideally use a framework like NSubstitute or Moq to fake the return of said call.
I'm currently myself at the stage where I've done my tinkering with discord.net and will soon start to code my bot properly, I will then write unit tests myself.
